cloud/openstack/os_server_group.py

```diff
@@ -0,0 +1,165 @@
+- os_server_group:
+    state: present
+    name: friends
+    policies:
+      - affinity
+
+# Delete a server group
+- os_server_group:
+    state: absent
+    name: friends
+'''
+
+RETURN = '''
+server_group:
+    description: Dictionary describing the server group.
+    returned: On success when I(state) is 'present'
+    type: dictionary
+    contains:
+        id:
+            description: Server Group ID
+            returned: success
+            type: string
+            sample: "cb4f45b4-08b6-4ea5-b4c7-6d39ec402e03"
+        name:
+            description: Server group
+            returned: success
+            type: string
+            sample: "security_group"
+        policies:
+            description: Policies applied to this group
+            returned: success
+            type: list
+            sample: "[affinity]"
+        members:
+            description: Hosts in this group
+            returned: success
+            type: list
+            sample: "[hostX]"
+        metadata:
+            description: Metadata for this group
+            returned: success
+            type: complex
+'''
+
+
+def _system_state_change(state, server_group):
+    if state == 'present' and not server_group:
+        return True
+    if state == 'absent' and server_group:
+        return True
+    return False
+
+
+def main():
+
+    argument_spec = openstack_full_argument_spec(
+        name=dict(required=True),
+        policies=dict(required=False, default=None, type='list'),
+        state=dict(default='present', choices=['absent', 'present']),
+    )
+
+    module_kwargs = openstack_module_kwargs()
+    module = AnsibleModule(
+        argument_spec,
+        required_if=([
+            ('state', 'present', ['policies'])
+        ]),
+        supports_check_mode=True,
+        **module_kwargs)
+
+    if not HAS_SHADE:
+        module.fail_json(msg='shade is required for this module')
+
+    name = module.params['name']
+    policies = module.params['policies']
+    state = module.params['state']
+
+    try:
+        cloud = shade.openstack_cloud(**module.params)
+        server_group = cloud.get_server_group(name)
+
+        if module.check_mode:
+            module.exit_json(changed=_system_state_change(state, server_group))
+
+        if state == 'present':
+            if server_group is None:
+                server_group = cloud.create_server_group(
+                    name=name, policies=policies)
+                changed = True
+            else:
+                changed = False
+            module.exit_json(changed=changed, server_group=server_group)
+        elif state == 'absent':
+            if server_group is None:
+                changed = False
+            else:
+                cloud.delete_server_group(server_group['id'])
+                changed = True
+            module.exit_json(changed=changed)
+
+    except shade.OpenStackCloudException as e:
+        module.fail_json(msg=str(e), extra_data=e.extra_data)
```

Comment:
The extra_data is not needed (old style message). That data will be included in the str(e) portion if it is present in the exception.